File: Awful.apk/src/main/java/com/ferg/awfulapp/AwfulApplication.java
package com.ferg.awfulapp;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ShortcutInfo;
import android.net.Uri;
import android.os.Build;
import android.os.StrictMode;
import android.webkit.WebView;

import com.ferg.awfulapp.announcements.AnnouncementsManager;
import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.network.NetworkUtils;
import com.ferg.awfulapp.preferences.AwfulPreferences;
import com.ferg.awfulapp.preferences.Keys;
import com.ferg.awfulapp.search.SearchFragment;
import com.ferg.awfulapp.sync.SyncManager;
import com.ferg.awfulapp.util.AwfulUtils;
import com.jakewharton.threetenabp.AndroidThreeTen;

import java.io.File;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

import androidx.core.content.pm.ShortcutInfoCompat;
import androidx.core.content.pm.ShortcutManagerCompat;
import androidx.core.graphics.drawable.IconCompat;
import timber.log.Timber;

public class AwfulApplication extends Application {
    private static final String APP_STATE_PREFERENCES = "app_state_prefs";
    /**
     * Used for storing misc app data, separate from user preferences, so onPreferenceChange callbacks aren't triggered
     */
    private static SharedPreferences appStatePrefs;

    /**
     * Stores the user agent used by web views in this application, which is required to be
     * initialised early on to correctly handle Cloudflare captcha situations.
     *
     * If the user is affected by Cloudflare captchas, the user-agents of all HTTP requests coming
     * from Awful must be synchronised. Setting a custom user-agent does not work, because
     * Cloudflare only allows mainstream browser user-agents.
     */
    private static String AWFUL_USER_AGENT = null;

    public static String getAwfulUserAgent() {
        return AWFUL_USER_AGENT;
    }

    /**
     * Instantiates an ephemeral WebView which is only used to retrieve the default user agent
     * of web views on this system.
     *
     * @return User-Agent string of WebView instances on this system.
     */
    private String getWebViewUserAgent() {
        if (Constants.DEBUG) {
           return "Microsoft Outlook 15.0.4833";
        }
        WebView view = new WebView(getApplicationContext());
        return view.getSettings().getUserAgentString();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        AWFUL_USER_AGENT = getWebViewUserAgent();

        // initialise the AwfulPreferences singleton first since a lot of things rely on it for a Context
        AwfulPreferences mPref = AwfulPreferences.getInstance(this);

        appStatePrefs = this.getSharedPreferences(APP_STATE_PREFERENCES, MODE_PRIVATE);
        mPref.setPreference(Keys.PROBATION_IGNORE, false);

        NetworkUtils.init(this);
        AndroidThreeTen.init(this);
        AnnouncementsManager.init();
        FontManager.createInstance(mPref, getAssets());
        updatePlatinumShortcuts(this);

        long hoursSinceInstall = getHoursSinceInstall();

        Timber.i("App installed %d hours ago", hoursSinceInstall);

        if (Constants.DEBUG) {
            Timber.d("*\n*\n*Debug active\n*\n*");
			/*
			This checks destroyed cursors aren't left open, and crashes (with a log) if it finds one
			Really this is here to avoid introducing any more leaks, since there are some issues with
			too many open cursors
			*/
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .penaltyLog()
                    .penaltyDeath()
                    .build());
        }

        SyncManager.sync(this);
    }

    private void updatePlatinumShortcuts(Context context) {
        if (!AwfulUtils.isAtLeast(Build.VERSION_CODES.N_MR1)) { return; }

        ArrayList<ShortcutInfoCompat> shortcuts = new ArrayList<>();
        if(!AwfulPreferences.getInstance().hasPlatinum) {
            ShortcutManagerCompat.setDynamicShortcuts(context, shortcuts);
            return;
        }
        ShortcutInfoCompat pms = new ShortcutInfoCompat.Builder(context, "pms")
                .setShortLabel(context.getResources().getString(R.string.private_message))
                .setLongLabel(context.getResources().getString(R.string.private_message))
                .setIcon(IconCompat.createWithResource(context, R.drawable.ic_inbox_black))
                .setIntent(new Intent(Intent.ACTION_VIEW, Uri.EMPTY,context, PrivateMessageActivity.class).putExtra("navigation event", "nav_show_private_messages"))
                .build();
        shortcuts.add(pms);

        ShortcutInfoCompat search = new ShortcutInfoCompat.Builder(context, "search")
                .setShortLabel(context.getResources().getString(R.string.search))
                .setLongLabel(context.getResources().getString(R.string.search))
                .setIcon(IconCompat.createWithResource(context, R.drawable.ic_search_black))
                .setIntent(BasicActivity.Companion.intentFor(SearchFragment.class, context, context.getString(R.string.search_forums_activity_title)).setAction(Intent.ACTION_VIEW))
                .build();
        shortcuts.add(search);
        ShortcutManagerCompat.setDynamicShortcuts(context, shortcuts);
    }

    /**
     * @return how long it's been since the app was updated
     */
    private long getHoursSinceInstall() {
        long hoursSinceInstall = Long.MAX_VALUE;
        try {
            PackageInfo packageInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            long millisSinceInstall = System.currentTimeMillis() - packageInfo.lastUpdateTime;
            hoursSinceInstall = TimeUnit.HOURS.convert(millisSinceInstall, TimeUnit.MILLISECONDS);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return hoursSinceInstall;
    }

    /**
     * Get the SharedPreferences used for storing basic app state.
     * <p>
     * These are separate from the default shared preferences, and won't trigger onPreferenceChange callbacks.
     *
     * @see AwfulPreferences.AwfulPreferenceUpdate#onPreferenceChange(AwfulPreferences, String)
     */
    public static SharedPreferences getAppStatePrefs() {
        return appStatePrefs;
    }

    @Override
    public File getCacheDir() {
        Timber.i("getCacheDir(): %s", super.getCacheDir());
        return super.getCacheDir();
    }

    @Override
    public void onTrimMemory(int level) {
        super.onTrimMemory(level);
        if (level != Application.TRIM_MEMORY_UI_HIDDEN && level != Application.TRIM_MEMORY_BACKGROUND) {
            NetworkUtils.clearImageCache();
        }
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        NetworkUtils.clearImageCache();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/AwfulApplication.kt
package com.ferg.awfulapp

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Typeface
import android.os.StrictMode
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.crashlytics.android.Crashlytics
import com.ferg.awfulapp.announcements.AnnouncementsManager
import com.ferg.awfulapp.network.NetworkUtils
import com.ferg.awfulapp.preferences.AwfulPreferences
import com.ferg.awfulapp.sync.SyncManager
import com.jakewharton.threetenabp.AndroidThreeTen
import io.fabric.sdk.android.Fabric
import timber.log.Timber
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

class AwfulApplication : Application(), AwfulPreferences.AwfulPreferenceUpdate {

    lateinit private var mPref: AwfulPreferences
    private val fonts = HashMap<String, Typeface>()
    private var currentFont: Typeface? = null

    val fontList: Array<String>
        get() {
            if (fonts.size == 0) {
                processFonts()
            }
            val keys = fonts.keys
            return keys.toTypedArray()
        }

    override fun onCreate() {
        super.onCreate()
        // initialise the AwfulPreferences singleton first since a lot of things rely on it for a Context
        mPref = AwfulPreferences.getInstance(this, this)
        appStatePrefs = this.getSharedPreferences(APP_STATE_PREFERENCES, Context.MODE_PRIVATE)


        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())

            StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .penaltyLog()
                    .penaltyDeath()
                    .build())
        } else {
            Fabric.with(this, Crashlytics())
            Timber.plant(CrashlyticsReportingTree())
            if (mPref.sendUsernameInReport) {
                Crashlytics.setUserName(mPref.username)
            }
        }

        NetworkUtils.init(this)
        AndroidThreeTen.init(this)
        AnnouncementsManager.init()
        onPreferenceChange(mPref, null)

        // work out how long it's been since the app was updated
        var hoursSinceInstall = java.lang.Long.MAX_VALUE
        try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            val millisSinceInstall = System.currentTimeMillis() - packageInfo.lastUpdateTime
            hoursSinceInstall = TimeUnit.HOURS.convert(millisSinceInstall, TimeUnit.MILLISECONDS)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        SyncManager.sync(this)
    }


    private fun setFontFromPreference(textView: TextView, baseFlags: Int) {
        var flags = baseFlags
        flags = if (flags < 0 && textView.typeface != null) {
            textView.typeface.style
        } else {
            Typeface.NORMAL
        }
        if (fonts.size == 0) {
            processFonts()
        }
        if (currentFont != null) {
            if (mPref.preferredFont.contains("mono")) {
                when (flags) {
                    Typeface.BOLD -> textView.setTypeface(currentFont, Typeface.BOLD)
                    Typeface.ITALIC -> textView.setTypeface(currentFont, Typeface.ITALIC)
                    Typeface.BOLD_ITALIC -> textView.setTypeface(currentFont, Typeface.BOLD_ITALIC)
                    Typeface.NORMAL -> textView.setTypeface(currentFont, Typeface.NORMAL)
                    else -> textView.setTypeface(currentFont, Typeface.NORMAL)
                }
            } else {
                textView.setTypeface(currentFont, flags)
            }
        }
    }

    private fun setFontFromPreferenceRecurse(viewGroup: ViewGroup, flags: Int) {
        for (i in 0 until viewGroup.childCount) {
            val child = viewGroup.getChildAt(i)
            if (child is TextView) {
                setFontFromPreference(child, flags)
            } else if (child is ViewGroup) {
                setFontFromPreferenceRecurse(child, flags)
            }
        }
    }

    fun setPreferredFont(view: View, flags: Int) {
        if (view is TextView) {
            setFontFromPreference(view, flags)
        } else if (view is ViewGroup) {
            setFontFromPreferenceRecurse(view, flags)
        }
    }

    override fun onPreferenceChange(prefs: AwfulPreferences?, key: String?) {
        currentFont = fonts[mPref.preferredFont]
        Timber.i("FONT SELECTED: %i", mPref.preferredFont)
    }

    private fun processFonts() {
        fonts.clear()
        fonts.put("default", Typeface.defaultFromStyle(Typeface.NORMAL))
        try {
            val files = assets.list("fonts")
            for (file in files) {
                val fileName = "fonts/" + file
                fonts.put(fileName, Typeface.createFromAsset(assets, fileName))
                Timber.v("Processed Font: $fileName")
            }
        } catch (e: IOException) {
            Timber.e(e)
        } catch (e: RuntimeException) {
            Timber.e(e)
        }

        onPreferenceChange(mPref, null)
    }

    override fun getCacheDir(): File {
        Timber.v("getCacheDir(): %s", super.getCacheDir())
        return super.getCacheDir()
    }


    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
        if (level != Application.TRIM_MEMORY_UI_HIDDEN && level != Application.TRIM_MEMORY_BACKGROUND) {
            NetworkUtils.clearImageCache()
        }
    }

    override fun onLowMemory() {
        super.onLowMemory()
        NetworkUtils.clearImageCache()
    }


    companion object {
        /**
         * Get the SharedPreferences used for storing basic app state.
         * These are separate from the default shared preferences, and won't trigger onPreferenceChange callbacks.
         * @see AwfulPreferences.AwfulPreferenceUpdate.onPreferenceChange
         */
        lateinit var appStatePrefs: SharedPreferences
        private val APP_STATE_PREFERENCES = "app_state_prefs"
    }
}
